### PR TITLE
CSS module: generated content

### PR DESCRIPTION
**Pull request by estelle**
Rewrote the two pages.

The module page now follows our new module structure.
the content page needed a rewrite: the values and browser support have changed a lot since this was originally written.

### CODE REVIEW

File: files/en-us/web/css/content/index.md [hunk 1]
```diff
@@ -62,43 +62,59 @@ content: unset;
...
+  - : The `<target>` data type includes three target functions, `<target-counter()>`, `<target-counters()>`, and `<target-text()>` that create cross-reference obtained from the target end of a link. See [Formal syntax](#formal_syntax).
 
-- [Accessibility support for CSS generated content – Tink](https://tink.uk/accessibility-support-for-css-generated-content/)
-- [Explanation of WCAG, Guideline 1.3 – MDN](/en-US/docs/Web/Accessibility/Understanding_WCAG/Perceivable#guideline_1.3_%e2%80%94_create_content_that_can_be_presented_in_different_ways)
-- [Understanding Success Criterion 1.3.1 | W3C Understanding WCAG 2.0](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html)
+- `<leader()>` {{Experimental_Inline}}
+
+  - : The `<leader()>` data type inclues a leader function: `leader( <leader-type> )`. This function accepts the keyword values `dotted`, `solid`, or `space` (equal to `leader(".")`, `leader("_")`, and `leader(" ")`, respectively), or a `<string>` as a parameter. When supported and used as a value for `content`, the leader-type provided will be inserted as a repeating pattern visually connecting content across a horizontal line.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:110]:
```suggestion
  - : The `<leader()>` data type inclues a leader function: `leader( <leader-type> )`. This function accepts the keyword values `dotted`, `solid`, or `space` (equal to `leader(".")`, `leader("_")`, and `leader(" ")`, respectively), or a `<string>` as a parameter. When supported and used as a value for `content`, the leader-type provided will be inserted as a repeating pattern, visually connecting content across a horizontal line.
```

File: files/en-us/web/css/content/index.md [hunk 2]
```diff
@@ -62,43 +62,59 @@ content: unset;
...
+
+- `attr(x)`
+
+  - : The `attr(x)` CSS function retrieves the value of an attribute of the selected element or pseudo-element's originating element. The value of the element's attribute `x` as an unparsed string. If there is no attribute `x`, an empty string is returned. The case-sensitivity of attribute name parameter depends on the document language.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:114]:
```suggestion
  - : The `attr(x)` CSS function retrieves the value of an attribute of the selected element, or the pseudo-element's originating element. The value of the element's attribute `x` is an unparsed string representing the attribute name. If there is no attribute `x`, an empty string is returned. The case sensitivity of the attribute name parameter depends on the document language.
```

File: files/en-us/web/css/content/index.md [hunk 3]
```diff
@@ -62,43 +62,59 @@ content: unset;
...
+
+- alternative text `/ <string> | <counter>`
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:116]:
```suggestion
- alternative text: `/ <string> | <counter>`
```

Not 100% sure if this is the best way, but when I originally read this it took me a while to get my head around the fact that the slash is part of the syntax. I thought some kind of separator might help.

File: files/en-us/web/css/content/index.md [hunk 4]
```diff
@@ -62,43 +62,59 @@ content: unset;
...
+  - : Alternative text may be specified for an image or any `<content-list>` items, by appending a forward slash and then a string of text or counter. The alternative text is intended for speech output by screen-readers, but may also be displayed in some browsers. Note that if the browser does not support alternative text, the `content` declaration will be considered invalid and will be ignored. The {{cssxref("string", "/ &lt;string>")}} or {{cssxref("counter", "/ &lt;counter>")}} specifies the "alt text" for the element.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:117]:
```suggestion
  - : Alternative text may be specified for an image or any `<content-list>` items, by appending a forward slash and then a string of text or a counter. The alternative text is intended for speech output by screen-readers, but may also be displayed in some browsers. Note that if the browser does not support alternative text, the `content` declaration will be considered invalid and will be ignored. The {{cssxref("string", "/ &lt;string>")}} or {{cssxref("counter", "/ &lt;counter>")}} data types specify the "alt text" for the element.
```

File: files/en-us/web/css/content/index.md [hunk 5]
```diff
@@ -110,14 +126,46 @@ CSS-generated content is not included in the [DOM](/en-US/docs/Web/API/Document_
 
 ## Examples
 
-### Headings and quotes
+The first examples created generated content on pseudo-elements. The last three are [examples of element replacement](#element-replacement-with-url).
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:129]:
```suggestion
The first five examples create generated content on pseudo-elements. The last three are [examples of element replacement](#element-replacement-with-url).
```

File: files/en-us/web/css/content/index.md [hunk 6]
```diff
@@ -110,14 +126,46 @@ CSS-generated content is not included in the [DOM](/en-US/docs/Web/API/Document_
...
 
-This example inserts quotation marks around quotes, and adds the word "Chapter" before headings.
+### Appending strings based on an element's class
+
+This example inserts generated text, coloring it red, after the text based on an element's class name.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:133]:
```suggestion
This example inserts generated text after the text of elements that have a particular class name. The text is colored red.
```

File: files/en-us/web/css/content/index.md [hunk 7]
```diff
@@ -154,142 +202,136 @@ q::before {
...
-                          rest of the content */
-}
 ```
 
 #### Result
 
-{{EmbedLiveSample('Headings_and_quotes', '100%', 200)}}
+{{EmbedLiveSample('Quotes', '100%', 200)}}
+
+Note the [type of quotes generated](/en-US/docs/Web/CSS/quotes#auto_quotes) is based on the language. Browsers add open- and close-quotes before and after {{HTMLElement("q")}} elements by default, so the quotes in this example would appear without them being explicitly set. They could have been turned off by setting `no-open-quote` and `no-close-quote`, or `none` value. They can also be turned off by setting the {{cssxref("quotes")}} property to `none` instead.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:211]:
```suggestion
Note the [type of quotes generated](/en-US/docs/Web/CSS/quotes#auto_quotes) is based on the language. Browsers add open- and close-quotes before and after {{HTMLElement("q")}} elements by default, so the quotes in this example would appear without them being explicitly set. They could have been turned off by setting the respective `content` property values to `no-open-quote` and `no-close-quote`, or by setting them both to `none`. They can also be turned off by setting the {{cssxref("quotes")}} property to `none` instead.
```

File: files/en-us/web/css/content/index.md [hunk 8]
```diff
@@ -154,142 +202,136 @@ q::before {
...
-{{EmbedLiveSample('Targeting_classes', '100%', 160)}}
+{{EmbedLiveSample('Strings_with_attribute_values', '100%', 200)}}
 
-### Images and element attributes
+The generated content is the value of the `href` attribute, prepended by "URL: ", with a space, all in parentheses.
 
-This example inserts an image before each link, and adds its `id` attribute after.
+### Adding an image with alternative text
+
+This example inserts an image before the link. Alternative text that a screen reader can output as speech in included,
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:288]:
```suggestion
This example inserts an image before the link. Alternative text that a screen reader can output as speech is included,
```

File: files/en-us/web/css/content/index.md [hunk 9]
```diff
@@ -154,142 +202,136 @@ q::before {
...
+which some browsers may display. To indicate if alternative text after a content list is supported or not, a fallback `content`, the CSS includes the message " - no alt text - ".
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:289]: This sentence is difficult to parse, and I'm not 100% sure what you are trying to say. Can you please clarify it?

File: files/en-us/web/css/content/index.md [hunk 10]
```diff
@@ -154,142 +202,136 @@ q::before {
...
 ```
 
 #### Result
 
-{{EmbedLiveSample('Images_and_element_attributes', '100%', 160)}}
+{{EmbedLiveSample('Adding_an_image_with_alternative_text', '100%', 60)}}
+
+If using a screen reader, it should speak the word "MOZILLA" when it reaches the image. If supported (if the "no alt text" is not showing), you can select the `::before` pseudo-element with your developer tools selection tool, and view the {{glossary("accessible name")}} in the accessibility panel.
+
+Note that on a browser that does not support the alternative text syntax, the whole line is invalid.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:324]:
```suggestion
Note that in browsers that don't support the alternative text syntax, the whole line is invalid.
```

Also, maybe put it in a note box if it is a note?

File: files/en-us/web/css/content/index.md [hunk 11]
```diff
@@ -154,142 +202,136 @@ q::before {
...
+In this case, the previous `content` value, if supported, will be used.
 
 ### Element replacement with `url()`
 
-This example replaces an element's content with an image {{cssxref("url", "url()")}}. Content added with `::before` or `::after` will not be generated as the contents of the element have been replaced.
+This example replaces a regular element! The element's contents are replaced with an SVG using the {{cssxref("url", "url()")}} image function. We include generated content, attempting to add the `id` of the element generated on `::after`. This pseudo-element is not generated in browsers that support content replacement as the element is replaced.
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:329]: Should "that support content replacement" be "that don't support content replacement"? The sentence sounds a bit contradictory as-is.

**Review by chrisdavidmills — CHANGES_REQUESTED** [review]
@estelle some fabulous work here; these are so much better than the originals. I've left you some suggestions to mull over.

File: files/en-us/web/css/css_generated_content/index.md [hunk 12]
```diff
@@ -7,9 +7,17 @@ spec-urls: https://drafts.csswg.org/css-content/
...
 
-**CSS Generated Content** is a module of CSS that defines how to add content to an element. Generated content can be used to add content to anonymous replaced elements or replace the content of a DOM node in very limited circumstances with a generated value.
+The **CSS generated content** module defines an element' content can be replaced and content can be added to a document with CSS.
 
-See the [how to guide for generated content](/en-US/docs/Learn/CSS/Howto/Generated_content) to learn more, and the {{cssxref("content")}} and {{cssxref("quotes")}} properties for implementation information.
+Generated content can be used for content replacement, in which case the content of a DOM node is replaced with a CSS `<image>`. The CSS generated content also enables generating language-specific quotes, create custom list item numbers and bullets, and visually adding content by generating content on select pseudo-elements as anonymous replaced elements.
+
+### Generated content in action
+
+{{EmbedGHLiveSample("css-examples/modules/generated_content.html", '100%', 720)}}
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/css_generated_content/index.md]: it does look bad in FF. I had increased the height b/c i thought that was the issue, but it wasn't. So, I made the height much smaller here and will address the css-example in a separate PR

File: files/en-us/web/css/css_generated_content/index.md [hunk 13]
```diff
@@ -18,11 +26,58 @@ See the [how to guide for generated content](/en-US/docs/Learn/CSS/Howto/Generat
...
+
+  - : Learn how to add text or image content to a document using the {{cssxref("content")}} property.
+
+- [Create fancy boxes with generated content](/en-US/docs/Learn/CSS/Howto/Create_fancy_boxes)
+
+  - : Example of styling generated content for visual effects.
+
+## Related concepts
+
+[CSS pseudo-elements](/en-US/docs/Web/CSS/CSS_pseudo-elements) module
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/css_generated_content/index.md]: made nested lists

File: files/en-us/web/css/content/index.md [hunk 14]
```diff
@@ -62,43 +62,59 @@ content: unset;
 
 ### Values
 
+The value can be one of two keywords — `none` or `normal` — or, for DOM nodes, the content replacement can be an `<image>`, otherwise a list of of one or more anonymous inline boxes, called a `<content-list>`, that will replace the content of the selected element in the specified order. Each content-list item is either `contents` or of type [`<string>`](#string), [`<image>`](#image), [`<counter>`](#counter), [`<quote>`](#quote), [`<target>`](#target), or [`<leader()>`](#leader), with an optional alternative text value of a `<string>` or `<counter>`:
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/content/index.md]: Changed to:
The value can be one of two keywords — `none` or `normal` — or either `<content-replacement>` when replacing a DOM node or a `<content-list>` for pseudo-elements and margin boxes, with optional alternative text value of a `<string>` or `<counter>`. Content replacement is always an `<image>`. A content-list is a list of of one or more anonymous inline boxes appearing in the order specified. Each content-list item is either `contents` or of type [`<string>`](#string), [`<image>`](#image), [`<counter>`](#counter), [`<quote>`](#quote), [`<target>`](#target), or [`<leader()>`](#leader):

File: files/en-us/web/css/content/index.md (hunk 10 above)
Comment (estelle) [reply files/en-us/web/css/content/index.md:324]: didn't put a note around, but did add an explanatory sentence.
> In browsers that don't support the alternative text syntax the whole declaration containing the alt text is invalid. In this case, the previous `content` value will be used, showing the image and "alt text is not supported" text.

File: files/en-us/web/css/content/index.md (hunk 11 above)
Comment (estelle) [reply files/en-us/web/css/content/index.md:327]: It's correct. Pseudo-elements aren't rendered on replaced elements. If the element is replaced, the ::after and ::before will not be generated. Tried to explain it better.

**Comment by estelle** [comment]
@chrisdavidmills Thank you for the review. i think i addressed all the comments.

File: files/en-us/web/css/css_generated_content/index.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (chrisdavidmills) [reply files/en-us/web/css/css_generated_content/index.md]: Sounds good.

File: files/en-us/web/css/css_generated_content/index.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (chrisdavidmills) [reply files/en-us/web/css/css_generated_content/index.md]: LGTM!

File: files/en-us/web/css/content/index.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (chrisdavidmills) [reply files/en-us/web/css/content/index.md]: This is an improvement, but it is still hard to parse and work out which of the keywords or data types relates to each of the three options you are talking about.

How about something like this:

```
The value can be:

  - One of two keywords — `none` or `normal`
  - `<content-replacement>` when replacing a DOM node. `<content-replacement>` is always an `<image>`.
  - A `<content-list>` when replacing pseudo-elements and margin boxes, with an optional alternative text value of a `<string>` or `<counter>`.  A content-list is a list of one or more anonymous inline boxes appearing in the order specified. Each `<content-list>` item is either `contents` or of type [`<string>`](#string), [`<image>`](#image), [`<counter>`](#counter), [`<quote>`](#quote), [`<target>`](#target), or [`<leader()>`](#leader).

The keywords and data types mentioned above are described in more detail below:

#### Simple keywords

<!-- List descriptions -->

#### Content replacement data types

<!-- List descriptions -->

####  Content list keywords and data types

<!-- List descriptions -->

```

?

File: files/en-us/web/css/content/index.md [hunk 15]
```diff
@@ -154,142 +202,136 @@ q::before {
...
-{{EmbedLiveSample('Targeting_classes', '100%', 160)}}
+{{EmbedLiveSample('Strings_with_attribute_values', '100%', 200)}}
 
-### Images and element attributes
+The generated content is the value of the `href` attribute, prepended by "URL: ", with a space, all in parentheses.
 
-This example inserts an image before each link, and adds its `id` attribute after.
+### Adding an image with alternative text
+
+This example inserts an image before all links. Two `content` values are provided. The later `content` value includes an image with alternative text that a screen reader can output as speech. If a browser does not support alternative text, this declaration will be considered invalid, with the previous `content` value displaying. This fallback content list include an image and the message " - alt text is not supported - ".
```
Comment (chrisdavidmills) [inline files/en-us/web/css/content/index.md:288]:
```suggestion
This example inserts an image before all links. Two `content` values are provided. The later `content` value includes an image with alternative text that a screen reader can output as speech. If a browser does not support alternative text, this declaration will be considered invalid, with the previous `content` value displaying. This fallback content list includes an image and the message " - alt text is not supported - ".
```

File: files/en-us/web/css/content/index.md (hunk 11 above)
Comment (chrisdavidmills) [reply files/en-us/web/css/content/index.md]: Yup, makes sense now

**Review by chrisdavidmills — CHANGES_REQUESTED** [review]
@estelle nearly there; thanks for all the hard work on this.

I just had one more suggested fix, and a further comment on the syntax section.

File: files/en-us/web/css/content/index.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/content/index.md]: Love it! @chrisdavidmills ready for re-review

File: files/en-us/web/css/css_generated_content/index.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/css_generated_content/index.md]: Fixed in https://github.com/mdn/css-examples/pull/152

**Review by chrisdavidmills — APPROVED** [review]
@estelle much better. Let's get this merged!
